### ISSUE TITLE
Unable to specify groupBy return type.

### ISSUE THREAD

**Issue by synthet1c** [selfrefactor/rambda#597]
Hey Dejan,

I'm trying to use groupBy in my code but have an incompatible type. Could I update the type signature to allow for specifying an input and an output?

This is a [simple example](https://stackblitz.com/edit/typescript-n6ux6q?devtoolsheight=33&file=index.ts) of what I'm dealing with.

```typescript
import { groupBy, prop } from 'rambda';

interface Thing {
  name: string;
  position: string;
}

interface Output {
  left: Thing[];
  right: Thing[];
}

const things = [
  { name: 'one', position: 'left' },
  { name: 'two', position: 'left' },
  { name: 'three', position: 'right' },
  { name: 'four', position: 'right' },
]

/*
Type '(list: Thing[]) => { [index: string]: Thing[]; }' is not assignable to type '(things: Thing[]) => Output'.
  Type '{ [index: string]: Thing[]; }' is missing the following properties from type 'Output': left, right(2322)
*/
const groupByPosition: (things: Thing[]) => Output = 
  groupBy<Thing>(prop('position'));

console.log(groupByPosition(things));
```

When I call groupByPosition the return type is `{ [index: string]: T[] }`. While it seem like it should be compatible with my `Output` interface it's displaying an error.

With the following type definition I can specify both the input and the output. Is this something I can make a pull request for?
```typescript
export function groupBy<T>(groupFn: (x: T) => string, list: T[]): { [index: string]: T[] };
export function groupBy<T>(groupFn: (x: T) => string): (list: T[]) => { [index: string]: T[] };
// <------------- New definitions -------------->
export function groupBy<T, U>(groupFn: (x: T) => string, list: T[]): U;
export function groupBy<T, U>(groupFn: (x: T) => string): (list: T[]) => U;
```

Then in my code I can use.
```typescript
const groupByPosition: (things: Thing[]) => Output = 
  groupBy<Thing, Output>(prop('position'));
```

If there is a better way to force the type can you let me know? I know I can change the signature of `Output` to have the non specific key but I don't want to make the code less specific if I can help it.

```typescript
interface Output {
  [key: string]: Thing[]
}
```

Cheers,

Andrew

**Comment by selfrefactor** [comment]
Thanks for the report Andrew. I take a look at your suggestion and I'll report back.

**Comment by selfrefactor** [comment]
I am currently in a sick leave, but afterwards I will look into the issue.

**Comment by synthet1c** [comment]
Hope it's not from covid or "the vaccine". Hope you feel better! Praying for you!

**Comment by selfrefactor** [comment]
I think it looks good. I will release it with the new version. I will write here once the release is published.

These are typings tests for your change:

```
import {groupBy} from 'rambda'
import {groupBy as groupByRamda, prop} from 'ramda'

interface Thing {
  name: string;
  position: string;
}

interface Output {
  left: Thing[];
  right: Thing[];
}

const things = [
  { name: 'one', position: 'left' },
  { name: 'two', position: 'left' },
  { name: 'three', position: 'right' },
  { name: 'four', position: 'right' },
]

describe('R.groupBy', () => {
  it('with one explicit types', () => {    
    const groupByPosition = groupBy<Thing>(prop('position'));

    // $ExpectError
    groupByRamda(prop('position'));
    
    const result = groupByPosition(things)
    result // $ExpectType { [index: string]: Thing[]; }
    result[9]  // $ExpectType Thing[]
    result.foo  // $ExpectType Thing[]
  })
  it('with two explicit types', () => {    
    const groupByPosition = groupBy<Thing, Output>(prop('position'));
    
    const result = groupByPosition(things)
    result // $ExpectType Output
  })
})

```

**Comment by selfrefactor** [comment]
I am closing the issue as the fix is released with version `7.0.0`.
